src/app/users/login.ts

```diff
@@ -0,0 +1,126 @@
+import {
+  getFailedLoginAttemptPerIpLimits,
+  getFailedLoginAttemptPerPhoneLimits,
+} from "@config/app"
+import { CouldNotFindUserFromPhoneError, UnknownRepositoryError } from "@domain/errors"
+import { RateLimitPrefix } from "@domain/rate-limit"
+import {
+  RateLimiterExceededError,
+  UserLoginIpRateLimiterExceededError,
+  UserLoginPhoneRateLimiterExceededError,
+} from "@domain/rate-limit/errors"
+import { createToken } from "@services/jwt"
+import { UsersRepository } from "@services/mongoose"
+import { PhoneCodesRepository } from "@services/mongoose/phone-code"
+import { TwilioClient } from "@services/twilio"
+import { RedisRateLimitService } from "@services/rate-limit"
+import { isTestAccountPhoneAndCode } from "."
+
+export const login = async ({
+  phone,
+  code,
+  logger,
+  ip,
+}: {
+  phone: PhoneNumber
+  code: PhoneCode
+  logger: Logger
+  ip: IpAddress
+}): Promise<JwtToken | ApplicationError> => {
+  const subLogger = logger.child({ topic: "login" })
+
+  {
+    const limitOk = await checkFailedLoginAttemptPerIpLimits(ip)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  {
+    const limitOk = await checkFailedLoginAttemptPerPhoneLimits(phone)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  // TODO:
+  // add fibonachi on failed login
+  // https://github.com/animir/node-rate-limiter-flexible/wiki/Overall-example#dynamic-block-duration
+
+  const validCode = await isCodeValid({ phone, code })
+  if (validCode instanceof Error) return validCode
+
+  await rewardFailedLoginAttemptPerIpLimits(ip)
+
+  const userRepo = UsersRepository()
+  let user: RepositoryError | User = await userRepo.findByPhone(phone)
+
+  if (user instanceof CouldNotFindUserFromPhoneError) {
+    subLogger.info({ phone }, "new user signup")
+
+    const userRaw = { phone } as NewUserInfo
+
+    const carrierInfo = await TwilioClient().getCarrier(phone)
+    if (carrierInfo instanceof Error) {
+      // non fatal error
+      subLogger.warn({ phone }, "impossible to fetch carrier")
+    } else {
+      userRaw.phoneMetadata = carrierInfo
+    }
+
+    user = await userRepo.persistNew(userRaw)
```

Comment:
Small nit, this could probably be `const persistedUser = ...` to reflect the difference between this variable and the `user` we fetch from the `findOne` higher up to help with keeping these concepts separate.

Reply:
they're the same concept? first one is the case the user exist, second is the case the user didn't exist because, but they both return `persistedUser`